frame/nft/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+		}
+	}
+
+	impl<T: Config> Inspect<AccountIdOf<T>> for Pallet<T> {
+		type ClassId = NFTClass;
+		type InstanceId = NFTInstanceId;
+
+		fn owner(class: &Self::ClassId, instance: &Self::InstanceId) -> Option<AccountIdOf<T>> {
+			Self::instance_owner((class, instance))
+		}
+
+		fn attribute(
+			class: &Self::ClassId,
+			instance: &Self::InstanceId,
+			key: &[u8],
+		) -> Option<Vec<u8>> {
+			Instance::<T>::get((class, instance)).and_then(|x| x.get(key).cloned())
+		}
+
+		fn class_attribute(class: &Self::ClassId, key: &[u8]) -> Option<Vec<u8>> {
+			Class::<T>::get(class).and_then(|x| x.get(key).cloned())
+		}
+	}
+
+	impl<T: Config> Create<AccountIdOf<T>> for Pallet<T> {
+		fn create_class(
+			class: &Self::ClassId,
+			who: &AccountIdOf<T>,
+			admin: &AccountIdOf<T>,
+		) -> DispatchResult {
+			ensure!(Self::class_owners(class).is_none(), Error::<T>::ClassAlreadyExists);
+			ClassOwners::<T>::insert(class, (who, admin));
+			Class::<T>::insert(class, BTreeMap::<Vec<u8>, Vec<u8>>::new());
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Transfer<AccountIdOf<T>> for Pallet<T> {
+		fn transfer(
+			class: &Self::ClassId,
+			instance: &Self::InstanceId,
+			destination: &AccountIdOf<T>,
+		) -> DispatchResult {
+			Self::ensure_instance_exists(class, instance)?;
+			InstanceOwner::<T>::insert((class, instance), destination);
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Mutate<AccountIdOf<T>> for Pallet<T> {
+		fn mint_into(
+			class: &Self::ClassId,
+			instance: &Self::InstanceId,
+			who: &AccountIdOf<T>,
+		) -> DispatchResult {
+			ensure!(
+				Self::instance_owner((class, instance)).is_none(),
+				Error::<T>::InstanceAlreadyExists
+			);
+			InstanceOwner::<T>::insert((class, instance), who);
+			Instance::<T>::insert((class, instance), BTreeMap::<Vec<u8>, Vec<u8>>::new());
+			Ok(())
+		}
+
+		fn burn_from(class: &Self::ClassId, instance: &Self::InstanceId) -> DispatchResult {
+			Self::ensure_instance_exists(class, instance)?;
+			InstanceOwner::<T>::remove((class, instance));
+			Instance::<T>::remove((class, instance));
+			Ok(())
+		}
+
+		fn set_attribute(
+			class: &Self::ClassId,
+			instance: &Self::InstanceId,
+			key: &[u8],
+			value: &[u8],
+		) -> DispatchResult {
+			Instance::<T>::try_mutate((class, instance), |entry| match entry {
+				Some(nft) => {
+					nft.insert(key.into(), value.into());
+					Ok(())
+				},
+				None => Err(Error::<T>::InstanceNotFound.into()),
+			})
+		}
+
+		fn set_typed_attribute<K: Encode, V: Encode>(
+			class: &Self::ClassId,
+			instance: &Self::InstanceId,
+			key: &K,
+			value: &V,
+		) -> DispatchResult {
+			key.using_encoded(|k| {
+				value.using_encoded(|v| Self::set_attribute(class, instance, k, v))
+			})
+		}
+
+		fn set_class_attribute(class: &Self::ClassId, key: &[u8], value: &[u8]) -> DispatchResult {
+			Self::ensure_class_exists(class)?;
+			Class::<T>::mutate(class, |entry| {
+				entry.as_mut().map(|nft_class| nft_class.insert(key.into(), value.into()));
+			});
+			Ok(())
```

Comment:
looks like this can rewrite using `try_mutate_exists` as here https://github.com/paritytech/substrate/blob/master/frame/assets/src/functions.rs#L782-L805

Reply:
How would you rewrite this? `try_mutate_exists` does mutate if Ok is returned, in this case we always mutate? I'm not sure I understand the example :thinking:

Reply:
Ah, you mean removing the `ensure_class_exists` and merge it in a `try_mutate` :+1: